spring-cloud-skipper-server-core/src/main/java/org/springframework/cloud/skipper/server/service/PackageMetadataService.java

```diff
@@ -50,10 +58,83 @@
 
 	private final RepositoryRepository repositoryRepository;
 
+	private final PackageMetadataRepository packageMetadataRepository;
+
+	private final ReleaseRepository releaseRepository;
+
 	private ResourceLoader resourceLoader;
 
-	public PackageMetadataService(RepositoryRepository repositoryRepository) {
+	public PackageMetadataService(RepositoryRepository repositoryRepository,
+			PackageMetadataRepository packageMetadataRepository,
+			ReleaseRepository releaseRepository) {
 		this.repositoryRepository = repositoryRepository;
+		this.packageMetadataRepository = packageMetadataRepository;
+		this.releaseRepository = releaseRepository;
+	}
+
+	@Transactional
+	public void deleteIfAllReleasesDeleted(String packageName) {
+		List<PackageMetadata> packageMetadataList = this.packageMetadataRepository.findByNameRequired(packageName);
+		for (PackageMetadata packageMetadata : packageMetadataList) {
+			deleteIfAllReleasesDeleted(packageMetadata);
+		}
+	}
+
+	/**
+	 * Delete all versions of the package metadata only if the latest releases currently using
+	 * it are in the StatusCode.DELETED state.
+	 * @param packageMetadata the package metadata
+	 */
+	@Transactional
+	public void deleteIfAllReleasesDeleted(PackageMetadata packageMetadata) {
+
+		List<Release> releases = this.releaseRepository.findByRepositoryIdAndPackageMetadataIdOrderByNameAscVersionDesc(
+				packageMetadata.getRepositoryId(),
+				packageMetadata.getId());
+		boolean canDelete = true;
+
+		List<Release> releasesFromLocalRepositories = filterReleasesFromLocalRepos(releases, packageMetadata.getName());
+
+		// Only keep latest release per release name.
+		Map<String, Release> latestReleaseMap = new HashMap<>();
+		for (Release release : releasesFromLocalRepositories) {
+			if (!latestReleaseMap.containsKey(release.getName())) {
+				latestReleaseMap.put(release.getName(), release);
+			}
+		}
+
+		// Find releases that are still 'active' so can't be deleted
+		List<String> activeReleaseNames = new ArrayList<>();
+		for (Release release : latestReleaseMap.values()) {
+			if (!release.getInfo().getStatus().getStatusCode().equals(StatusCode.DELETED)) {
+				canDelete = false;
+				activeReleaseNames.add(release.getName());
+			}
+		}
+		if (!canDelete) {
+			Repository repository = this.repositoryRepository.findOne(packageMetadata.getRepositoryId());
+			throw new SkipperException(String.format("Can not delete Package Metadata [%s:%s] in Repository [%s]. " +
```

Comment:
Given we iterate over the list of all the package metadata by the given `name`, throwing exception for the specific package metadata here (when the condition is met) would break the execution for the rest.

Reply:
I don't understand the comment.  

The method argument takes a specific packageMetadata.  We iterate overall the releases in local repositories and see if any have a status different than DELETED.  If so, then I think we can just abort any attempt to delete anything, until all releases are in the DELETED state.

Reply:
I meant the iteration done [here](https://github.com/spring-cloud/spring-cloud-skipper/pull/447/files/8ad87dc11106938a5cc130c0b6d3838ff4db632b#diff-2aa49f674c378b62d111b7898ed5526aR77) which subsequently calls this method.